### PR TITLE
ENH Support type hinting using `mypy`

### PR DESCRIPTION
**Pull request by illumination-k**
#### Reference Issue or PRs

Please see #39

#### What does your PR implement? Be specific.

This PR support type hinting by `mypy` and fixed some docstring which incompatible with `mypy` type checking.

### CODE REVIEW

**Comment by BorisMuzellec** [comment]
Hi @illumination-k, thanks a lot for your work!

This is our first contribution from an external collaborator, we're very excited about it :).

I have just approved the CI to run on this PR. It seems that some checks have failed, could you have a look at it if you have time?

FYI I'll be away next week so your interlocutor will probably be @maikia.

Thanks again for your help, hope to merge this soon!

**Comment by illumination-k** [comment]
Sorry for failing CI. I ran CI and pre-commit scripts manually and maybe OK. If failed, I will check tomorrow.

**Comment by maikia** [comment]
Hi @illumination-k
Thanks for your contribution! I think accidentally you added `docs/auto_examples` (they are generated by sphinx when `make html` is run and should not be a part of git). Can you remove them and I will review your PR?

**Review by maikia — COMMENTED** [review]
@illumination-k thanks again for your contribution. Very nice and through PR. 

Could you provide me with the details on your settings of mypy, or even better if you can add it to the CI (and I will reproduce the steps locally). 
ATM if I run naive mypy it fails with errors, eg
image
To my knowledge those errors can also be coming from the version of Python. The one I am using locally (as well as the one used in our CI) is `3.8`.

I also didn’t realise before but `api/docstrings` should also be removed from the PR, and yes, adding this as well as auto examples to `.gitignore` is a great idea. Feel free to open a new PR if you like.

**Review by maikia — APPROVED** [review]
@illumination-k thanks again for your PR! I extended it to have mypy running within CI and corrected a bug coming from one of the changes in the code. Can you please make sure that you agree with all the changes I did?

@BorisMuzellec @mandreux-owkin can one of you review before merge (as I committed to this PR I prefer that there is someone else checking if all is good, thx)

File: pydeseq2/dds.py [hunk 1]
```diff
@@ -266,13 +271,22 @@ def fit_genewise_dispersions(self):
 
         # Exclude genes with all zeroes
         non_zero = ~(self.counts == 0).all()
-        self.non_zero_genes = non_zero[non_zero].index
+        _non_zero_genes = non_zero[non_zero].index
+
+        if isinstance(_non_zero_genes, pd.MultiIndex):
+            raise ValueError("Must not multi index")
```
Comment (BorisMuzellec) [inline pydeseq2/dds.py:277]:
```suggestion
            raise ValueError("MultiIndexes are not allowed")
```

File: pydeseq2/dds.py [hunk 2]
```diff
@@ -598,19 +612,20 @@ def calculate_cooks(self):
 
         num_vars = self.design_matrix.shape[1]
         # Keep only non-zero genes
-        normed_counts = self.counts.loc[:, self.non_zero_genes].div(self.size_factors, 0)
+        none_zero_gene_counts: pd.DataFrame = self.counts.loc[:, self.non_zero_genes]
+        normed_counts = none_zero_gene_counts.div(self.size_factors, 0)
```
Comment (BorisMuzellec) [inline pydeseq2/dds.py:616]:
```suggestion
        non_zero_gene_counts: pd.DataFrame = self.counts.loc[:, self.non_zero_genes]
        normed_counts = non_zero_gene_counts.div(self.size_factors, 0)
```

File: pydeseq2/dds.py [hunk 3]
```diff
@@ -598,19 +612,20 @@ def calculate_cooks(self):
...
+
         dispersions = robust_method_of_moments_disp(normed_counts, self.design_matrix)
         V = self._mu_LFC + dispersions * self._mu_LFC**2
-        squared_pearson_res = (
-            self.counts.loc[:, self.non_zero_genes] - self._mu_LFC
-        ) ** 2 / V
+
+        squared_pearson_res = (none_zero_gene_counts - self._mu_LFC) ** 2 / V
```
Comment (BorisMuzellec) [inline pydeseq2/dds.py:621]:
```suggestion
        squared_pearson_res = (non_zero_gene_counts - self._mu_LFC) ** 2 / V
```

File: pydeseq2/utils.py [hunk 4]
```diff
@@ -397,15 +410,16 @@ def irls_solver(
...
-
+    # if type(counts) == pd.Series:
+    #     counts = counts.values
+    # if type(size_factors) == pd.Series:
+    #     size_factors = size_factors.values
+    # if type(design_matrix) == pd.Series:
+    #     X = design_matrix.values
+    # else:
+    #     X = design_matrix
+
```
Comment (BorisMuzellec) [inline pydeseq2/utils.py:421]: If this commented block of code is not useful, it should be deleted

**Review by BorisMuzellec — CHANGES_REQUESTED** [review]
LGTM, up to minor suggestions. 

Thanks a lot @illumination-k (and @maikia) for the work!

**Comment by BorisMuzellec** [comment]
Hi, now that the AnnData integration has been merged, there will be numerous merge conflicts to settle. I'll handle them, if that's OK for you.

**Comment by illumination-k** [comment]
Sure! Sorry for my late response, @BorisMuzellec .

**Review by BorisMuzellec — APPROVED** [review]

**Comment by illumination-k** [comment]
Your commit maybe inactivated type hinting. There are many `Any` types. I think that in order to fully benefit from type hints with mypy, you should use a command such as mypy . --check-untyped-defs --strict to perform type checking.

**Comment by BorisMuzellec** [comment]
Hi @illumination-k, indeed, thanks for checking!

It seems like I made some git manipulation mistakes, sorry for that... 

I don't think it's necessary to close this PR. I'll just revert to where it was before I made changes and take it from there. We'll squash anyways at merge time.

File: docs/source/conf.py [hunk 5]
```diff
@@ -62,37 +62,24 @@
     "python": ("https://docs.python.org/3", None),
     "numpy": ("https://numpy.org/doc/stable/", None),
     "pandas": ("https://pandas.pydata.org/docs/", None),
-    "anndata": ("https://anndata.readthedocs.io/en/latest/", None),
```
Comment (maikia) [inline docs/source/conf.py]: is this related to this PR?

File: pydeseq2/utils.py [hunk 6]
```diff
@@ -114,7 +120,7 @@ def test_valid_counts(counts_df):
     if counts_df.isna().any().any():
         raise ValueError("NaNs are not allowed in the count matrix.")
     if ~counts_df.apply(
-        lambda s: pd.to_numeric(s, errors="coerce").notnull().all()
+        lambda s: pd.to_numeric(s, errors="coerce").notnull().all()  # type: ignore
```
Comment (maikia) [inline pydeseq2/utils.py:123]: as much as with imports type: ignore is self explanatory here I think a comment should be added on why it is used

File: pydeseq2/utils.py [hunk 7]
```diff
@@ -397,15 +410,16 @@ def irls_solver(
...
-    if type(counts) == pd.Series:
-        counts = counts.values
-    if type(size_factors) == pd.Series:
-        size_factors = size_factors.values
-    if type(design_matrix) == pd.Series:
-        X = design_matrix.values
-    else:
-        X = design_matrix
-
+    # if type(counts) == pd.Series:
```
Comment (maikia) [inline pydeseq2/utils.py:413]: shouldn't those comments be removed?

File: pydeseq2/utils.py [hunk 8]
```diff
@@ -549,6 +562,7 @@ def fit_alpha_mle(
     assert optimizer in ["BFGS", "L-BFGS-B"]
 
     if prior_reg:
+        # Note: assertion is not work when using numpy
```
Comment (maikia) [inline pydeseq2/utils.py:565]:
```suggestion
        # Note: assertion is not working when using numpy
```

**Review by maikia — APPROVED** [review]
This PR did grow quite a bit. LGTM. just few minor comments. 
@illumination-k can you also comment if you agree with all the changes done to your PR?

**Comment by BorisMuzellec** [comment]
> @BorisMuzellec there are also some conflicts needing resolve

Yes, I am currently handling those. These are due to the anndata merge

**Comment by illumination-k** [comment]
I would prefer to address `mypy . --check-untyped-def` and ignore untyped things coming from `anndata` integration in this PR. Because there are many untyped things and it is pretty tough work. I'm not so familiar with anndata, but it seems not to support type stubs, so maybe some work is also need to handle anndata type. 

We should address anndata typing and `--strict` flag in the next PRs.

**Comment by BorisMuzellec** [comment]
Hi @illumination-k and @maikia, I did my best to pass the the `mypy . --check-untyped-defs` (without the `strict` option).
Could you have a look?

I agree with @illumination-k that we should deal with more intricate typing issues brought by the `anndata` integration in a new PR.

**Comment by maikia** [comment]
@BorisMuzellec I left some comments in my previous review which were still not addressed. Otherwise I agree with your last comment

File: docs/source/conf.py (hunk 5 above)
Comment (BorisMuzellec) [reply docs/source/conf.py]: This line was restored by the new merge

File: pydeseq2/utils.py (hunk 7 above)
Comment (BorisMuzellec) [reply pydeseq2/utils.py]: Indeed, done

File: pydeseq2/utils.py (hunk 6 above)
Comment (BorisMuzellec) [reply pydeseq2/utils.py:123]: Indeed. As I am personally not so familiar with `mypy`, @illumination-k could you add a quick comment to explain? Thanks!

Comment (illumination-k) [reply pydeseq2/utils.py:123]: This is my mistake. I just put this as temporary for ignore, and I forgot to resolve it.

Comment (BorisMuzellec) [reply pydeseq2/utils.py:123]: Ok, thanks for clarifying :)

**Comment by BorisMuzellec** [comment]
@illumination-k and @maikia since all comments are now resolve, unless there is something else you would like to change I'll merge this PR today
